src/click/core.py

```diff
@@ -1014,25 +1014,26 @@ def get_help_option_names(self, ctx: Context) -> list[str]:
         return list(all_names)
 
     def get_help_option(self, ctx: Context) -> Option | None:
-        """Returns the help option object."""
-        help_options = self.get_help_option_names(ctx)
+        """Returns the help option object.
 
-        if not help_options or not self.add_help_option:
+        Unless ``add_help_option`` is ``False``.
+        """
+        help_option_names = self.get_help_option_names(ctx)
+
+        if not help_option_names or not self.add_help_option:
             return None
 
-        def show_help(ctx: Context, param: Parameter, value: str) -> None:
-            if value and not ctx.resilient_parsing:
-                echo(ctx.get_help(), color=ctx.color)
-                ctx.exit()
-
-        return Option(
-            help_options,
-            is_flag=True,
-            is_eager=True,
-            expose_value=False,
-            callback=show_help,
-            help=_("Show this message and exit."),
-        )
+        # Avoid circular import.
+        from .decorators import help_option
+
+        def dummy_func() -> None:
+            pass
+
+        # Call @help_option decorator to produce an help option with proper
+        # defaults and attach it to the dummy function defined above.
+        help_option(*help_option_names)(dummy_func)
+
+        return dummy_func.__click_params__.pop()  # type: ignore[no-any-return, attr-defined]
```

Comment:
A throw-away function isn't needed here. The option decorator can be applied to the command `self` and popped from `self.params`.

```suggestion
        # apply help_option decorator and pop resulting option
        help_option(*help_option_names)(self)
        return self.params.pop() # type: ignore[return-value]
```